RealmSwift/SwiftUI.swift

```diff
@@ -1450,59 +1449,68 @@ private class ObservableAsyncOpenStorage: ObservableObject {
         case loggedIn(User)
         case loggedOut
     }
-    private var appState: AppState {
-        didSet {
-            switch appState {
-            case .loggedIn(let user):
-                self.asyncOpenForUser(user)
-            case .loggedOut:
-                asyncOpenState = .waitingForUser
-            }
-        }
-    }
+    private var appState: AppState = .loggedOut
 
     // Cancellables
     private var appCancellable = [AnyCancellable]()
     private var asyncOpenCancellable = [AnyCancellable]()
 
-    @Published var asyncOpenState: AsyncOpenState = .connecting {
-        willSet {
-            objectWillChange.send()
-        }
-    }
+    @Published fileprivate var asyncOpenState: AsyncOpenState
 
-    fileprivate func update(_ partitionValue: PartitionValue?, _ configuration: Realm.Configuration) {
-        var open = false
-        if let partitionValue = partitionValue {
-            let bsonValue = AnyBSON(partitionValue: partitionValue)
-            if self.partitionValue != bsonValue {
-                self.partitionValue = bsonValue
-                open = true
-            }
-        }
+    init(asyncOpenKind: AsyncOpenKind, app: App, configuration: Realm.Configuration?, partitionValue: AnyBSON?) {
+        self.asyncOpenKind = asyncOpenKind
+        self.app = app
+        self.configuration = configuration
+        self.partitionValue = partitionValue
 
-        // We don't want to use the `defaultConfiguration` from the environment, we only want to use this environment value in @AsyncOpen if is not the default one
-        if configuration != .defaultConfiguration, self.configuration != configuration {
-            if let partitionValue = configuration.syncConfiguration?.partitionValue {
-                self.partitionValue = partitionValue
-            }
-            self.configuration = configuration
-            open = true
-        }
-        if open {
-            self.asyncOpen()
+        // Initialising the state value depending on the user status, before first rendering.
+        if let user = app.currentUser {
+            appState = .loggedIn(user)
+            asyncOpenState = .connecting
+        } else {
+            asyncOpenState = .waitingForUser
         }
     }
 
-    private func asyncOpen() {
-        if case let .loggedIn(user) = appState {
+    var setupHasRun = false
+    func setup() {
+        guard !setupHasRun else { return }
+        initAsyncOpen()
+        setupHasRun = true
+    }
+
+    private func initAsyncOpen() {
+        if case .loggedIn(let user) = appState {
+            // we only open the realm on initialisation if there is a user logged.
             asyncOpenForUser(user)
         }
+
+        // we observe the changes in the app state to check for user changes,
+        // we store an internal state, so we could react to those changes (user login, user change, logout).
+        app.objectWillChange.sink { [weak self] app in
+            switch self?.appState {
```

Comment:
Should do `guard let self = self else { return }` rather than checking for nil on every use.